src/server/shell/decoration/basic_manager.h

```diff
@@ -47,27 +48,31 @@ class BasicManager :
     public Manager
 {
 public:
-    using DecorationBuilder = std::function<std::unique_ptr<Decoration>(
-        std::shared_ptr<shell::Shell> const& shell,
-        std::shared_ptr<scene::Surface> const& surface)>;
-
     BasicManager(
         mir::ObserverRegistrar<mir::graphics::DisplayConfigurationObserver>& display_configuration_observers,
-        DecorationBuilder&& decoration_builder);
+        std::shared_ptr<mir::graphics::GraphicBufferAllocator> const& buffer_allocator,
+        std::shared_ptr<Executor> const& executor,
+        std::shared_ptr<input::CursorImages> const& cursor_images);
+
     ~BasicManager();
 
     void init(std::weak_ptr<shell::Shell> const& shell) override;
     void decorate(std::shared_ptr<scene::Surface> const& surface) override;
     void undecorate(std::shared_ptr<scene::Surface> const& surface) override;
     void undecorate_all() override;
 
+protected:
+    std::unordered_map<scene::Surface*, std::unique_ptr<Decoration>> decorations;
+
```

Comment:
I'm not enthusiastic about making member data public (`protected` is public in disguise).

In this case the only use is in tests, which manipulate `decorations` directly. That seems fragile and suggests there's an abstraction missing.

I'll try to come up with a concrete suggestion.